trajectory_executor/src/trajectory_executor/trajectory_executor.cpp

```diff
@@ -0,0 +1,188 @@
+    {
+        // Hard coded stub for MVP since plugin manager won't be developed yet
+        // TODO: Query plugin manager to receive actual list of plugins and their corresponding topics
+        ROS_DEBUG("Executing stub behavior for plugin discovery MVP...");
+        std::map<std::string, std::string> out;
+
+        std::string default_control_plugin;
+        _default_nh->param<std::string>("default_control_plugin", default_control_plugin, "NULL");
+
+        std::string default_control_plugin_topic;
+        _default_nh->param<std::string>("default_control_plugin_topic", default_control_plugin_topic, "NULL");
+
+        out[default_control_plugin] = default_control_plugin_topic;
+        return out;
+    }
+    
+    void TrajectoryExecutor::onNewTrajectoryPlan(cav_msgs::TrajectoryPlan msg)
+    {
+        std::unique_lock<std::mutex> lock(_cur_traj_mutex); // Acquire lock until end of this function scope
+        ROS_DEBUG("Received new trajectory plan!");
+        ROS_DEBUG_STREAM("New Trajectory plan ID: " << msg.trajectory_id);
+        ROS_DEBUG_STREAM("New plan contains " << msg.trajectory_points.size() << " points");
+
+        _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(msg));
+        ROS_DEBUG_STREAM("Successfully swapped trajectories!");
+
+        // TODO: Force traj emit
+        // TODO: Reset timer state
+    }
+
+    void TrajectoryExecutor::onTrajEmitTick(const ros::TimerEvent& te)
+    {
+        std::unique_lock<std::mutex> lock(_cur_traj_mutex);
+        ROS_DEBUG("TrajectoryExecutor tick start!");
+        
+
+        if (_cur_traj != nullptr) {
+            if (_timesteps_since_last_traj > 0) {
+                _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(trimFirstPoint(*_cur_traj)));
+            }
+            if (!_cur_traj->trajectory_points.empty()) {
+                std::string control_plugin = _cur_traj->trajectory_points[0].controller_plugin_name;
+                std::map<std::string, ros::Publisher>::iterator it = _traj_publisher_map.find(control_plugin);
+                if (it != _traj_publisher_map.end()) {
+                    ROS_DEBUG("Found match for control plugin %s at point %d in current trajectory!",
+                        control_plugin.c_str(),
+                        _timesteps_since_last_traj);
+                    it->second.publish(*_cur_traj);
+                } else {
+                    std::ostringstream description_builder;
+                    description_builder << "No match found for control plugin " 
+                        << control_plugin << " at point " 
+                        << _timesteps_since_last_traj << " in current trajectory!";
+                    shutdown(description_builder.str());
+                }
+                _timesteps_since_last_traj++;
+            } else {
+                shutdown("Ran out of trajectory data to consume!");
```

Comment:
I am not 100% sure. But I think those shutdown function can be replaced by the exception handling function in CARMANodeHandle.

Reply:
Fixed!